apps/web/src/middleware.ts

```diff
@@ -1,15 +1,21 @@
+import { NextResponse } from 'next/server'
 import { shouldGeoBlock } from '@pancakeswap/utils/geoBlock'
-import { NextRequest, NextResponse } from 'next/server'
+import { withContext } from './context'
 
-export async function middleware(req: NextRequest) {
-  const res = NextResponse.next()
+export default withContext('userIp', (setContext, req) => {
+  let ip = req.ip ?? req.headers.get('x-real-ip')
+  const forwardedFor = req.headers.get('x-forwarded-for')
+  if (!ip && forwardedFor) {
+    ip = forwardedFor.split(',').at(0) ?? 'Unknown'
+  }
 
+  // determine if user should see feature
+  setContext('userIp', ip ?? 'ip-unavailable')
   if (shouldGeoBlock(req.geo)) {
     return NextResponse.redirect(new URL('/451', req.url))
   }
-
-  return res
-}
+  return NextResponse.next()
```

Comment:
Let's make it more elegant. I'm thinking something like

```typescript
// src/middleware.ts
export const middleware = withMiddlewares(
  [middleware1, middleware2],
  defaultHandler, // Should be optional
)

// src/middlewares/index.ts
export function withMiddlewares(middlewares, defaultHandler) {
}

// src/middlewares/geoBlock.ts
// handle geo block

// src/middlewares/webNotification.ts
// handle notification ab testing

// src/middlewares/context.ts
// for the context, I think we only need to handle cookies for now. As we're not using response header.
// We can name all cookies set from middleware as pancake_m_*, e.g. pancake_m_notification_beta
```

Reply:
let's not pass the ip back to the client. We can process it in our middleware

Reply:
we can have an extra middleware `src/middlewares/userIp.ts` to extract ip from request headers